README.md
# Labels for Pivotal Tracker

This Chrome extension for Pivotal Tracker allows you to set custom colors for
label names.

![](https://lh3.googleusercontent.com/KQcEsgjs5-CwNuLorWlhRy8fiSBWrjVuyEU1RYzxDtLTQmg7I8HuJ8oqQ_hSrwAHJ9qmUTAKsA=s1280-h800-e365-rw)

## Installation

Install the exension here:

https://chrome.google.com/webstore/detail/pivotal-labels/njkofaephafobbeabnmmmijchhhdomio

Then navigate to the "options" for the extension to start adding labels.

## License

MIT


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.0.3] - 2017-04-04
### Fixed
- Injected style was being applied to non-custom labels

## [0.0.2] - 2017-04-04
### Changed
- Automatically lowercase name before saving
- Add a description to labels section
- Only clear fields if the save is successful
- Updated README

## [0.0.1] - 2017-04-03
### Added
- Initial commit


src/inject/inject.js
function textColor(hex) {
  let r = parseInt(hex.slice(1,3), 16)
  let g = parseInt(hex.slice(3,5), 16)
  let b = parseInt(hex.slice(5,7), 16)
  let a = Math.pow((Math.pow(0.299*r, 2) + Math.pow(0.587*g, 2) + Math.pow(0.114*b, 2)), 0.5)
  if (a < 128) {
    return "#FFFFFF"
  } else {
    return "#000000"
  }
}

chrome.extension.sendMessage({}, function(response) {
  var readyStateCheckInterval = setInterval(function() {
    if (document.readyState === "complete") {
      clearInterval(readyStateCheckInterval);

      var observer = new MutationObserver(function (mutations) {
        mutations.forEach(handleMutationEvents);
      });

      // configuration of the observer:
      var config = {
        attributes: true,
        characterData: true,
        childList: true,
        subtree: true
      };

      observer.observe(document, config);

      var handleMutationEvents = function handleMutationEvents(mutation) {
        Array.prototype.forEach.call(mutation.addedNodes, styleLabelsInNode);
        styleLabelsInNode(mutation.target);
      }

      var styleLabelsInNode = function styleLabelsInNode(node) {
        if (nodeIsElement(node)) {
          styleLabels(findLabelsInNode(node));
        }
      }

      var nodeIsElement = function nodeIsElement(node) {
        return (typeof node.querySelectorAll !== 'undefined');
      }

      var findLabelsInNode = function findLabelsInNode(node) {
        return node.querySelectorAll('a.label');
      }

      var styleLabels = function styleLabels(labels) {
        chrome.storage.sync.get({'labels': {}}, function(storage) {
          Array.prototype.forEach.call(labels, function(label) {
            var labelText = label.textContent.replace(/\, $/g, '');
            var labelColor = storage.labels[labelText];
            if (labelColor) {
              var hex = "#" + labelColor
              label.style.setProperty('background-color', hex);
              label.style.setProperty('color', textColor(hex));
              label.textContent = labelText;
              label.classList.add('custom')
            } else {
              label.style.backgroundColor = '';
              label.classList.remove('custom')
            }
          });
        });
      }
    }
  }, 10);
});


options.js
function textColor(hex) {
  let r = parseInt(hex.slice(1,3), 16)
  let g = parseInt(hex.slice(3,5), 16)
  let b = parseInt(hex.slice(5,7), 16)
  let a = Math.pow((Math.pow(0.299*r, 2) + Math.pow(0.587*g, 2) + Math.pow(0.114*b, 2)), 0.5)
  if (a < 128) {
    return "#FFFFFF"
  } else {
    return "#000000"
  }
}

function setColor(hex) {
  document.getElementById('labelColor').value = hex.substring(1, 7);
}

function empty(obj) {
  return Object.getOwnPropertyNames(obj).length == 0
}

function set_status(message) {
  var status = document.getElementById('status');
  status.textContent = message;
  setTimeout(function() {
    status.textContent = '';
  }, 750);
}

function delete_label(labelName) {
  chrome.storage.sync.get({'labels': {}}, function(storage) {
    delete storage.labels[labelName]
    chrome.storage.sync.set(storage, function() {
      update_labels(storage)
    });
  });
}

function update_labels(storage) {
  if (!empty(storage.labels)) {
    document.getElementById('labelHeader').className = '';
    document.getElementById('labelDescription').className = '';
  } else {
    document.getElementById('labelHeader').className = 'hidden';
    document.getElementById('labelDescription').className = 'hidden';
  }

  let labelDiv = document.getElementById('labels');
  labelDiv.innerText = ''
  for (let key in storage.labels) {
    let label = document.createElement("a")
    label.appendChild(document.createTextNode(key));
    label.className = 'label custom'
    let hex = '#' + storage.labels[key]
    label.style.backgroundColor = hex
    label.style.color = textColor(hex)
    let del = document.createElement("a");
    del.appendChild(document.createTextNode("Delete"))
    del.href = "#"
    del.onclick = () => delete_label(key)
    let div = document.createElement("div");
    div.className = 'preview'
    div.appendChild(label);
    div.appendChild(del);
    labelDiv.appendChild(div);
  }
}

// Saves options to chrome.storage
function save_options() {
  var labelName = document.getElementById('labelName').value.toLowerCase();
  var labelColor = document.getElementById('labelColor').value;
  chrome.storage.sync.get({'labels': {}}, function(storage) {
    if (labelName && labelColor) {
      storage.labels[labelName] = labelColor
      chrome.storage.sync.set(storage, function() {
        document.getElementById('labelName').value = '';
        document.getElementById('labelColor').value = '';
        update_labels(storage);
      });
    } else {
      set_status('Missing fields!')
    }
  });
}

// Restores select box and checkbox state using the preferences
// stored in chrome.storage.
function restore_options() {
  chrome.storage.sync.get({'labels': {}}, update_labels);
  var palette = document.getElementById('palette')
  var colors = [
    "#e7e7e7", "#b6cff5", "#98d7e4", "#e3d7ff", "#fbd3e0", "#f2b2a8",
    "#c2c2c2", "#4986e7", "#2da2bb", "#b99aff", "#f691b2", "#fb4c2f",
    "#ffc8af", "#ffdeb5", "#fbe983", "#fdedc1", "#b3efd3", "#a2dcc1",
    "#ff7537", "#ffad46", "#cca6ac", "#ebdbde", "#42d692", "#16a765",
  ]
  colors.forEach(function(hex) {
    let color = document.createElement("div");
    color.className = 'color';
    color.style.backgroundColor = hex
    color.style.color = textColor(hex)
    color.textContent = 'a'
    color.onclick = () => setColor(hex);
    palette.appendChild(color);
  });
}
document.addEventListener('DOMContentLoaded', restore_options);
document.getElementById('save').addEventListener('click',
    save_options);
